File: palpng/pangapp.js

const fs = require("fs");

const PangPalette = require("./palette");
const PangPage = require("./page");
const PangDownloadable = require("./downloadable");
const PangResolution = require("./resolution");

const PangTemplate = require("./template");



function PangApp () {
	this.preferences = JSON.parse(fs.readFileSync("preferences.json"));
	this.paths = this.preferences.paths;
	this.gaText = fs.readFileSync(this.paths.analytics, 'utf8');
	this.alphabet = "abcdefghijklmnopqrstuvwxyz";
	this.nl = "<!-- \n -->";

	this.setupPalettes();
	this.setupPages();
	this.setupDownloadables();
	this.setupResolutions();
	this.setupTemplates();
}

PangApp.prototype.setupPalettes = function () {
	this.paletteJson = JSON.parse(fs.readFileSync(this.paths.palettes));

	this.palettes = [];
	for (let paletteJson of this.paletteJson.palettes) {
		let palette = new PangPalette(this, paletteJson);
		this.palettes.push(palette);
	}
	this.numberOfPalettes = this.palettes.length;
}

PangApp.prototype.setupPages = function () {
	this.pageLength = this.preferences.pageLength;
	this.numberOfPages = Math.ceil(this.numberOfPalettes / this.pageLength);
	this.pages = [];
	for (let pageNumber = 1; pageNumber <= this.numberOfPages; pageNumber++) {
		let page = new PangPage(this, pageNumber);
		this.pages.push(page);
	}
}

PangApp.prototype.setupDownloadables = function () {
	this.downloadables = [];
	for (let json of this.preferences.downloadables) {
		let downloadable = new PangDownloadable(this, json);
		this.downloadables.push(downloadable);
	}
}

PangApp.prototype.setupResolutions = function () {
	this.argsJson = JSON.parse(fs.readFileSync(this.paths.argsJson));
	this.resolutions = [];
	for (let json of this.argsJson.resolutions) {
		let resolution = new PangResolution(this, json);
		this.resolutions.push(resolution);
	}
}

PangApp.prototype.setupTemplates = function () {
	this.templates = {
		page: new PangTemplate(this, "page"),
		palette: new PangTemplate(this, "palette")
	};
}



PangApp.prototype.getLastPage = function () {
	if (this.pages.length) {
		return this.pages[this.pages.length - 1];
	}
	return null;
}

PangApp.prototype.getLastPalette = function () {
	if (this.palettes.length) {
		return this.palettes[this.palettes.length - 1];
	}
	return null;
}



PangApp.prototype.getNumberOfPages = function () {
	return this.pages.length;
}

PangApp.prototype.getNumberOfPalettes = function () {
	return this.palettes.length;
}

PangApp.prototype.getNumberOfDownloadables = function () {
	return this.downloadables.length;
}

PangApp.prototype.getNumberOfResolutions = function () {
	return this.resolutions.length;
}



PangApp.prototype.savePageCss = function () {
	for (let page of this.pages) {
		page.saveCss();
	}
}

PangApp.prototype.savePaletteCss = function () {
	for (let palette of this.palettes) {
		palette.saveCss();
	}
}

PangApp.prototype.saveCss = function () {
	this.savePageCss();
	this.savePaletteCss();
}



PangApp.prototype.savePageHtml = function () {
	for (let page of this.pages) {
		page.saveHtml();
	}
}

PangApp.prototype.savePaletteHtml = function () {
	for (let palette of this.palettes) {
		palette.saveHtml();
		// break; // while I test
	}
}

PangApp.prototype.saveHtml = function () {
	this.savePageHtml();
	this.savePaletteHtml();
}



PangApp.prototype.consoleLog = function () {
	console.log(`\t(${this.getNumberOfPages()} pages)`);
	console.log(`\t(${this.getNumberOfPalettes()} palettes)`);
	console.log(`\t(${this.getNumberOfDownloadables()} downloadables)`);
	console.log(`\t(${this.getNumberOfResolutions()} resolutions)`);
}



module.exports = PangApp;

File: cpp/pangapp.cpp
#include "pang/pangapp.hpp"

#include <fstream>

#include "pang/palette.hpp"
#include "pang/resolution.hpp"



pang::App::App ()
{
	auto args_path = "args.json";
	if (!fs::exists(args_path))
	{
		std::cout << "Not found: " << args_path << "\n";
		return;
	}

	std::ifstream args_fs(args_path);
	this->args_json = nj::json::parse(args_fs);

	for (auto rj : this->args_json["resolutions"]) {
		auto resolution = new pang::Resolution(this, rj);
		this->resolutions.push_back(resolution);
	}

	auto paletts_path = this->args_json["paths"]["palettes"];
	std::ifstream paletts_fs(paletts_path);
	this->palettes_json = nj::json::parse(paletts_fs);

	for (auto pj : this->palettes_json["palettes"]) {
		auto palette = new pang::Palette(this, pj);
		this->palettes.push_back(palette);
	}
}

pang::App::~App ()
{
	for (auto palette : this->palettes)
	{
		delete palette;
	}

	for (auto resolution : this->resolutions)
	{
		delete resolution;
	}
}



void pang::App::producePngs ()
{
	for (auto palette : this->palettes)
	{
		for (auto resolution : this->resolutions)
		{
			palette->producePngs(resolution);
			break;
		}
		break;
	}
}

void pang::App::consoleLog ()
{
	std::cout << "(" << this->palettes.size() << " palettes)\n";
}



